Pock/Widgets/NowPlaying/NowPlayingHelper.swift

```diff
@@ -70,19 +70,200 @@ class NowPlayingHelper {
+                        
+                        // Download the artwork
+                        self.artworkDownloadTask = URLSession.shared.dataTask(with: url, completionHandler: { (data, response, error) in
+                            
+                            if error != nil {
+                                completionHandler(nil)
+                                return
+                            }
+
+                            DispatchQueue.main.async {
+                                self.cachedAlbumArtName = search
+                                self.cachedAlbumArt = data
+                                // Set the artwork to the image
+                                completionHandler(self.cachedAlbumArt)
+                            }
+                        })
+                            
+                        self.artworkDownloadTask!.resume()
+                    }
+                    else {
+                        completionHandler(nil)
+                    }
+                }
+            }
+        }
+        
+        artworkAPITask!.resume()
+    }
+    
+    
+    
+    var albumArtFallbackTimer: Timer? = nil
+    
+    var lastUIUpdateNanos: UInt64? = nil
+    
+    // flag to indicate whether to fallback to network icon
+    var timesLeftTryUpdatingMediaContentManually: Int = 2
+
+    
+    @objc func fireTimerIconFailed() {
+        albumArtFallbackTimer = nil
+        if (self.nowPlayingItem.image == nil) {
+            self.updateArtwork(search: "\(self.nowPlayingItem.title ?? "") \(self.nowPlayingItem.artist ?? "")".replacingOccurrences(of: " ", with: "+").addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)!) { image in
+                self.nowPlayingItem.image = image
+                NotificationCenter.default.post(name: NowPlayingHelper.kNowPlayingItemDidChange, object: nil)
+            }
+        }
+    }
+    
+    // added this variable because `updateMediaContent` is called many times with different information
+    var lastNowPlayingItem: NowPlayingItem?
+    
+    func updateUI(withTimer: Bool) {
+        self.lastUIUpdateNanos = DispatchTime.now().uptimeNanoseconds
+        NotificationCenter.default.post(name: NowPlayingHelper.kNowPlayingItemDidChange, object: nil)
+        if withTimer {
+            if self.albumArtFallbackTimer == nil {
+                // fallback to network artwork after 20 seconds
+                DispatchQueue.main.async {
+                    if (self.timesLeftTryUpdatingMediaContentManually <= 0) {
+                        // we have tried getting the artwork with official api manuall but it didn't work, get it from the iTunes API
+                        self.timesLeftTryUpdatingMediaContentManually = 2
+                        self.albumArtFallbackTimer = Timer.scheduledTimer(timeInterval: 20.0, target: self, selector: #selector(NowPlayingHelper.fireTimerIconFailed), userInfo: nil, repeats: false)
```

Comment:
You could simplify the code by setting the fallback artwork first, and then once the iTunes API returns the image could be updated.

Reply:
The iTunes API is the least good solution because it can make mistakes when searching by artist and song name.
The native player API sometimes doesn't automatically notify the program when the artwork has been loaded, that's why I check for it manually first after 1.5 secs and then after 8 secs. When this fails the iTunes API is used after 5 secs.